src/shared/sleep-config.c

```diff
@@ -207,67 +314,180 @@ static bool battery_discharge_rate_is_valid(int battery_discharge_rate) {
+                                        usec_t after_timestamp) {
+
+        _cleanup_(sd_device_enumerator_unrefp) sd_device_enumerator *e = NULL;
+        sd_device *dev;
+        int battery_last_capacity, battery_current_capacity, battery_discharge_rate, r;
+
+        assert(last_capacity);
+        assert(current_capacity);
+
+        r = battery_enumerator_new(&e);
+        if (r < 0)
+                return log_debug_errno(r, "Failed to initialize battery enumerator: %m");
+
+        FOREACH_DEVICE(e, dev) {
+                const char *battery_name;
+                uint64_t system_hash_id;
+
+                r = sd_device_get_property_value(dev, "POWER_SUPPLY_NAME", &battery_name);
+                if (r < 0) {
+                        log_device_debug_errno(dev, r, "Failed to read battery name: %m");
+                        continue;
+                }
+
+                battery_last_capacity = PTR_TO_INT(hashmap_get(last_capacity, battery_name));
+                battery_current_capacity = PTR_TO_INT(hashmap_get(current_capacity, battery_name));
+                if (battery_current_capacity >= battery_last_capacity) {
+                        log_debug("Battery was not discharged during suspension");
+                        continue;
+                }
+                battery_discharge_rate = (battery_last_capacity - battery_current_capacity) * USEC_PER_HOUR / (after_timestamp - before_timestamp);
+                r = get_system_battery_identifier_hash(dev, &system_hash_id);
+                if (r < 0)
+                        return log_debug_errno(r, "Failed to generate system battery identifier hash: %m");
+
+                r = put_battery_discharge_rate(battery_discharge_rate, system_hash_id);
+                if (r < 0)
+                        log_warning_errno(r, "Failed to update battery discharge rate, ignoring: %m");
+        }
+
+        return 0;
+}
+
+/* calculate the suspend interval for each battery and then return the sum of it */
+
+int get_total_suspend_interval(Hashmap *last_capacity, usec_t *ret) {
+        _cleanup_(sd_device_enumerator_unrefp) sd_device_enumerator *e = NULL;
+        usec_t suspend_interval = 0, total_suspend_interval = 0;
+        int battery_last_capacity = 0, r;
+        sd_device *dev;
+
+        assert(last_capacity);
+
+        r = battery_enumerator_new(&e);
+        if (r < 0)
+                return log_debug_errno(r, "Failed to initialize battery enumerator: %m");
+
+        FOREACH_DEVICE(e, dev) {
+                const char *battery_name;
+                int previous_discharge_rate = 0;
+
+                r = sd_device_get_property_value(dev, "POWER_SUPPLY_NAME", &battery_name);
+                if (r < 0) {
+                        log_device_debug_errno(dev, r, "Failed to read battery name: %m");
+                        continue;
+                }
+
+                battery_last_capacity = PTR_TO_INT(hashmap_get(last_capacity, battery_name));
```

Comment:
the item might not be in the hashmap, this needs to handle that possibility

Reply:
added a check. leaving this unresolved. let me know if that needs change.